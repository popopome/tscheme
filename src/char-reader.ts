export interface ICharPos {
  pos: number;
  line: number;
  col: number;
}
export interface ICharReader {
  position(): ICharPos;
  current(): string;
  next(): string;
  seek(pos: number): void;
  getString(begin: ICharPos, end: ICharPos): string;
  getLineString(line: number): string;
}
export function makeCharReader(code: string): ICharReader {
  let pos = -1;
  let lines: number[] = [0];
  let col = 0;

  return {
    getLineString(line: number): string {
      if (line < lines.length) {
        const begin = lines[line];
        let end = line + 1 < lines.length ? lines[line + 1] : -1;
        if (end === -1) {
          end = begin;
          while (end < code.length) {
            if (code[end] === '\n') break;
            ++end;
          }
        }
        return code.substring(begin, end);
      }
      return '';
    },
    position() {
      return { pos, line: lines.length - 1, col };
    },
    current() {
      return code[pos];
    },
    next() {
      ++pos;
      if (pos >= code.length) return '';
      updateLineCol(pos);
      return this.current();
    },
    seek(position: number) {
      pos = position;
      for (let i = 0; i < lines.length; ++i) {
        if (pos < lines[i]) {
          lines.splice(0, i);
          updateLineCol(pos);
        }
      }
    },
    getString(begin: ICharPos, end: ICharPos) {
      return code.substring(begin.pos, end.pos);
    }
  };

  function updateLineCol(pos: number) {
    if (pos === 0) {
      lines = [0];
      col = 0;
    } else if (code[pos - 1] === '\n') {
      lines.push(pos);
      col = 0;
    } else {
      col = pos - lines[lines.length - 1];
    }
  }
}
