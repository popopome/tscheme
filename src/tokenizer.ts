import { makeCharReader, ICharReader, ICharPos } from './char-reader';

const WS_CHARS = new Set(' \t\n\r'.split(''));
const DIGITS = new Set('0123456789'.split(''));

function isWhitespace(c: string) {
  return WS_CHARS.has(c);
}
function isDigit(c: string) {
  return DIGITS.has(c);
}

function consumeWhitespace(reader: ICharReader) {
  let c;
  while ((c = reader.next())) {
    if (!isWhitespace(c)) return;
  }
}

export enum TokenType {
  EOF = 'EOF',
  OPEN_PAREN = 'OPEN_PAR',
  CLOSE_PAREN = 'CLOSE_PAREN',
  STRING = 'STRING',
  WORD = 'WORD',
  SYMBOL = 'SYMBOL',
  NUMBER = 'NUMBER',
  LIST = 'LIST'
}

export interface Token {
  type: TokenType;
  begin: ICharPos;
  end: ICharPos;
  value: string | number;
}
export interface Tokenizer {
  next(): Token;
  current(): Token;
}

function makeToken(
  type: TokenType,
  begin: ICharPos,
  end: ICharPos,
  value: string | number
): Token {
  return {
    type,
    begin,
    end,
    value
  };
}

function raiseParseError(msg: string, pos: ICharPos) {
  const err = new Error('parse error');
  (err as any).pos = pos;
  (err as any).msg = msg;
  throw err;
}

function trimDoubleQuote(s: string) {
  return s.substring(1, s.length - 1);
}

function consumeStringToken(reader: ICharReader): any {
  const begin = reader.position();
  let prev = reader.current();

  let c;
  while ((c = reader.next())) {
    if (c === '"') {
      // Escape?
      if (prev !== '\\') {
        reader.next();
        return makeToken(
          TokenType.STRING,
          begin,
          reader.position(),
          trimDoubleQuote(reader.getString(begin, reader.position()))
        );
      }
    }

    prev = c;
  }

  raiseParseError(`String should be closed with "`, reader.position());
}

function parseNumber(s: string): number | undefined {
  const v = parseInt(s);
  if ('' + v === s) return v;
  const r = parseFloat(s);
  if ('' + r === s) return r;
  return undefined;
}

function consumeNextToken(reader: ICharReader): any {
  const begin = reader.position();
  let c;
  while ((c = reader.next())) {
    if (isWhitespace(c) || c === '(' || c === ')') break;
  }

  const value = reader.getString(begin, reader.position());
  const parsedValue = parseNumber(value);
  return makeToken(
    parsedValue === undefined ? TokenType.SYMBOL : TokenType.NUMBER,
    begin,
    reader.position(),
    parsedValue === undefined ? value : parsedValue
  );
}

export function makeTokenizer(reader: ICharReader): Tokenizer {
  reader.next();
  let curr: Token;

  return {
    current(): Token {
      return curr;
    },
    next(): Token {
      return (curr = internalNext());
    }
  };
  function internalNext(): Token {
    const c = reader.current();
    if (!c) {
      return makeToken(TokenType.EOF, reader.position(), reader.position(), '');
    }

    let token;
    if (isWhitespace(c)) {
      consumeWhitespace(reader);
      return internalNext();
    }

    if (c === '(') {
      const pos = reader.position();
      reader.next();
      return makeToken(TokenType.OPEN_PAREN, pos, reader.position(), '(');
    }
    if (c === ')') {
      const pos = reader.position();
      reader.next();
      return makeToken(TokenType.CLOSE_PAREN, pos, reader.position(), ')');
    }
    if (c === '"') {
      return consumeStringToken(reader);
    }

    return consumeNextToken(reader);
  }
}
