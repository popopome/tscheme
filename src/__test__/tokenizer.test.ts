import { makeTokenizer, TokenType } from '../tokenizer';
import { makeCharReader } from '../char-reader';

test('tokenizer should parse empty list well', () => {
  const t = makeTokenizer(makeCharReader('(a)'));
  expect(t.next().type).toEqual(TokenType.OPEN_PAREN);
  expect(t.next().type).toEqual(TokenType.SYMBOL);
  expect(t.next().type).toEqual(TokenType.CLOSE_PAREN);
  expect(t.next().type).toEqual(TokenType.EOF);
});

test('tokenizer should make valid tokens', () => {
  const t = makeTokenizer(makeCharReader('( a b c "d1 d2 d3" 123 f)'));
  expect(t.next()).toMatchSnapshot();
  expect(t.next()).toMatchSnapshot();
  expect(t.next()).toMatchSnapshot();
  expect(t.next()).toMatchSnapshot();
  expect(t.next()).toMatchSnapshot();
  expect(t.next()).toMatchSnapshot();
});

test('tokenizer should report error', () => {
  try {
    const t = makeTokenizer(makeCharReader('( "thank'));
    t.next();
    t.next();
  } catch (err) {
    expect(JSON.stringify(err)).toMatchSnapshot();
    return;
  }
  expect(false).toBe('SHOULD NOT BE HERE');
});
