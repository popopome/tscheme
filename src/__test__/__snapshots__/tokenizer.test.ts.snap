// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer should make valid tokens 1`] = `
Object {
  "begin": Object {
    "col": 0,
    "line": 0,
    "pos": 0,
  },
  "end": Object {
    "col": 1,
    "line": 0,
    "pos": 1,
  },
  "type": "OPEN_PAR",
  "value": "(",
}
`;

exports[`tokenizer should make valid tokens 2`] = `
Object {
  "begin": Object {
    "col": 2,
    "line": 0,
    "pos": 2,
  },
  "end": Object {
    "col": 3,
    "line": 0,
    "pos": 3,
  },
  "type": "SYMBOL",
  "value": "a",
}
`;

exports[`tokenizer should make valid tokens 3`] = `
Object {
  "begin": Object {
    "col": 4,
    "line": 0,
    "pos": 4,
  },
  "end": Object {
    "col": 5,
    "line": 0,
    "pos": 5,
  },
  "type": "SYMBOL",
  "value": "b",
}
`;

exports[`tokenizer should make valid tokens 4`] = `
Object {
  "begin": Object {
    "col": 6,
    "line": 0,
    "pos": 6,
  },
  "end": Object {
    "col": 7,
    "line": 0,
    "pos": 7,
  },
  "type": "SYMBOL",
  "value": "c",
}
`;

exports[`tokenizer should make valid tokens 5`] = `
Object {
  "begin": Object {
    "col": 8,
    "line": 0,
    "pos": 8,
  },
  "end": Object {
    "col": 18,
    "line": 0,
    "pos": 18,
  },
  "type": "STRING",
  "value": "d1 d2 d3",
}
`;

exports[`tokenizer should make valid tokens 6`] = `
Object {
  "begin": Object {
    "col": 19,
    "line": 0,
    "pos": 19,
  },
  "end": Object {
    "col": 22,
    "line": 0,
    "pos": 22,
  },
  "type": "NUMBER",
  "value": 123,
}
`;

exports[`tokenizer should report error 1`] = `"{\\"pos\\":{\\"pos\\":8,\\"line\\":0,\\"col\\":7},\\"msg\\":\\"String should be closed with \\\\\\"\\"}"`;
